File: examples/go/chatcore.go
// Package main — crypto core for the Go chat-xdk example bot.
//
// ChatCore is a thin, network-free wrapper around the chatxdk binding.
// Everything that touches the SDK lives here so it can be unit-tested directly
// (see chatcore_test.go). The four core feature touchpoints are all here:
//
//   - key management     -> LoadKeys / GenerateAndRegister / SetIdentity / SetSigningKeys
//   - conversation keys  -> PrepareConversationKeyChange / DecryptConversationKey
//     (plus the SDK's opt-in key cache, enabled in NewChatCore)
//   - message encryption -> EncryptReply
//   - event decryption   -> DecryptBatch (DecryptEvents) and DecryptOne (DecryptEvent)
package main

import (
	"encoding/base64"

	"github.com/xdevplatform/chat-xdk/go/chatxdk"
)

// ChatCore wraps a single unlocked chatxdk.Chat for one bot identity.
type ChatCore struct {
	chat              *chatxdk.Chat
	SigningKeyVersion string
}

// NewChatCore creates an empty (locked) core with the SDK's conversation-key
// cache enabled, so encrypt calls resolve the key from previously decrypted
// key-change events instead of the caller threading it through.
func NewChatCore() *ChatCore {
	chat := chatxdk.New()
	chat.SetCacheKeys(true)
	return &ChatCore{chat: chat, SigningKeyVersion: "1"}
}

// Close releases native resources.
func (c *ChatCore) Close() { c.chat.Close() }

// -- Key management ---------------------------------------------------------

// LoadKeys imports an existing base64 private-key blob (identity[+signing])
// registered under signingKeyVersion.
func (c *ChatCore) LoadKeys(privateKeysB64, signingKeyVersion string) error {
	keys, err := base64.StdEncoding.DecodeString(privateKeysB64)
	if err != nil {
		return err
	}
	if err := c.chat.ImportKeysWithVersion(keys, signingKeyVersion); err != nil {
		return err
	}
	c.SigningKeyVersion = signingKeyVersion
	return nil
}

// SetIdentity sets the session identity once the bot's user id is known;
// every encrypt/prepare call after this resolves its sender from the session
// instead of taking it as an argument.
func (c *ChatCore) SetIdentity(userID string) error {
	return c.chat.SetIdentity(userID, c.SigningKeyVersion)
}

// SetSigningKeys stores the senders' signing keys for decrypt calls that
// pass nil signing keys. Each call replaces the previous set.
func (c *ChatCore) SetSigningKeys(keys []chatxdk.SigningKeyEntry) error {
	return c.chat.SetSigningKeys(keys)
}

// GenerateAndRegister creates a fresh identity. Returns the registration
// payload to POST to the X API plus the exported private blob (base64 for
// persistence; the SDK returns raw bytes).
func (c *ChatCore) GenerateAndRegister() (payload *chatxdk.PublicKeyRegistrationPayload, privateKeysB64 string, err error) {
	payload, err = c.chat.GenerateKeypairs()
	if err != nil {
		return nil, "", err
	}
	privateKeys, err := c.chat.ExportKeys()
	if err != nil {
		return nil, "", err
	}
	return payload, base64.StdEncoding.EncodeToString(privateKeys), nil
}

// PublicKeys returns the loaded identity + signing public keys.
func (c *ChatCore) PublicKeys() (*chatxdk.PublicKeys, error) {
	return c.chat.GetPublicKeys()
}

// -- Conversation keys ------------------------------------------------------

// PrepareConversationKeyChange generates, encrypts, and signs a conversation-key
// change under the session identity. Leave conversationID empty for a
// one-to-one to derive it; pass it for a group.
func (c *ChatCore) PrepareConversationKeyChange(publicKeys []chatxdk.PublicKeyInput, conversationID string) (*chatxdk.PreparedConversationChange, error) {
	return c.chat.PrepareConversationKeyChange(chatxdk.ConversationKeyChangeParams{
		PublicKeys:     publicKeys,
		ConversationID: conversationID,
	})
}

// DecryptConversationKey ECIES-decrypts one conversation key (base64 in, raw bytes out).
func (c *ChatCore) DecryptConversationKey(encryptedKeyB64 string) ([]byte, error) {
	return c.chat.DecryptConversationKey(encryptedKeyB64)
}

// -- Decryption: the two paths ---------------------------------------------

// DecryptBatch is the batch path used on initial conversation load. It
// extracts conversation keys from KeyChange events (caching them for the
// encrypt calls), and decrypts every message. Passing nil signingKeys uses
// the set stored via SetSigningKeys.
func (c *ChatCore) DecryptBatch(eventsB64 []string, signingKeys []chatxdk.SigningKeyEntry) (*chatxdk.DecryptEventsResult, error) {
	return c.chat.DecryptEvents(eventsB64, signingKeys)
}

// DecryptOne is the single-event path used for each new event after the
// initial load. Passing nil maps resolves conversation keys from the SDK's
// key cache and signing keys from the set stored via SetSigningKeys;
// explicit maps override both.
func (c *ChatCore) DecryptOne(eventB64 string, conversationKeys map[string][]byte, signingKeys []chatxdk.SigningKeyEntry) (*chatxdk.Event, error) {
	return c.chat.DecryptEvent(eventB64, conversationKeys, signingKeys)
}

// -- Message encryption -----------------------------------------------------

// SendBody holds the fields the X API expects for an encrypted message.
type SendBody struct {
	MessageID                    string `json:"message_id"`
	EncodedMessageCreateEvent    string `json:"encoded_message_create_event"`
	EncodedMessageEventSignature string `json:"encoded_message_event_signature"`
	ConversationToken            string `json:"conversation_token,omitempty"`
}

// ReplyOptions are the optional send parameters for EncryptReply.
type ReplyOptions struct {
	// ReplyToEvent is the raw base64 event being replied to; when set the
	// message becomes a threaded reply and the SDK derives (and embeds) the
	// reply preview from the raw event.
	ReplyToEvent string
	// ReplyToCkces are raw base64 key-change events needed to decrypt the
	// original when it used a different key version than this reply.
	ReplyToCkces []string
	// Entities are (start, end, type) byte ranges within the text.
	Entities []chatxdk.EntityTuple
	// Attachments are attachment descriptors (e.g. a media reference).
	Attachments []chatxdk.AttachmentDescriptor
	// TTLMsec makes the message disappear after the given lifetime (0 = none).
	TTLMsec int64
	// ConversationKey/ConversationKeyVersion override the SDK's cached key —
	// needed only when the key never arrived in a decrypted event (e.g. a
	// just-created group).
	ConversationKey        []byte
	ConversationKeyVersion string
}

// EncryptReply encrypts + signs a message under the session identity and
// returns the send body. The conversation key resolves from the SDK's key
// cache unless opts overrides it. Without opts.ReplyToEvent (opts may be
// nil) it sends a fresh message via EncryptMessage; with it, the SDK's
// EncryptReply builds a *threaded* reply.
func (c *ChatCore) EncryptReply(conversationID, text string, opts *ReplyOptions) (*SendBody, error) {
	if opts == nil {
		opts = &ReplyOptions{}
	}
	var ttl *int64
	if opts.TTLMsec > 0 {
		ttl = &opts.TTLMsec
	}
	var payload *chatxdk.SendPayload
	var err error
	if opts.ReplyToEvent == "" {
		payload, err = c.chat.EncryptMessage(chatxdk.EncryptMessageParams{
			ConversationID:         conversationID,
			Text:                   text,
			ConversationKey:        opts.ConversationKey,
			ConversationKeyVersion: opts.ConversationKeyVersion,
			Entities:               opts.Entities,
			Attachments:            opts.Attachments,
			TTLMsec:                ttl,
		})
	} else {
		payload, err = c.chat.EncryptReply(chatxdk.EncryptReplyParams{
			ConversationID:         conversationID,
			Text:                   text,
			ReplyToEvent:           opts.ReplyToEvent,
			ReplyToCkces:           opts.ReplyToCkces,
			ConversationKey:        opts.ConversationKey,
			ConversationKeyVersion: opts.ConversationKeyVersion,
			Entities:               opts.Entities,
			Attachments:            opts.Attachments,
			TTLMsec:                ttl,
		})
	}
	if err != nil {
		return nil, err
	}
	return &SendBody{
		// The SDK generates the message id and returns it in the payload.
		MessageID:                    payload.MessageID,
		EncodedMessageCreateEvent:    payload.EncryptedContent,
		EncodedMessageEventSignature: payload.EncodedEventSignature,
	}, nil
}

// EncryptAddReaction encrypts + signs a reaction add targeting a raw event.
// conversationKey may be nil to use the SDK's cached key.
func (c *ChatCore) EncryptAddReaction(targetEventB64, emoji string, conversationKey []byte, conversationKeyVersion string) (*SendBody, error) {
	return c.encryptReaction(true, targetEventB64, emoji, conversationKey, conversationKeyVersion)
}

// EncryptRemoveReaction encrypts + signs a reaction remove targeting a raw
// event. conversationKey may be nil to use the SDK's cached key.
func (c *ChatCore) EncryptRemoveReaction(targetEventB64, emoji string, conversationKey []byte, conversationKeyVersion string) (*SendBody, error) {
	return c.encryptReaction(false, targetEventB64, emoji, conversationKey, conversationKeyVersion)
}

func (c *ChatCore) encryptReaction(add bool, targetEventB64, emoji string, conversationKey []byte, conversationKeyVersion string) (*SendBody, error) {
	// The conversation id and target sequence id derive from the raw event.
	params := chatxdk.EncryptReactionParams{
		TargetEvent:            targetEventB64,
		Emoji:                  emoji,
		ConversationKey:        conversationKey,
		ConversationKeyVersion: conversationKeyVersion,
	}
	encrypt := c.chat.EncryptAddReaction
	if !add {
		encrypt = c.chat.EncryptRemoveReaction
	}
	payload, err := encrypt(params)
	if err != nil {
		return nil, err
	}
	return &SendBody{
		// The SDK generates the message id and returns it in the payload.
		MessageID:                    payload.MessageID,
		EncodedMessageCreateEvent:    payload.EncryptedContent,
		EncodedMessageEventSignature: payload.EncodedEventSignature,
	}, nil
}

// -- Group management ---------------------------------------------------------

// PrepareGroupCreate prepares a group creation under the session identity:
// fresh key + the two required signatures.
func (c *ChatCore) PrepareGroupCreate(publicKeys []chatxdk.PublicKeyInput, conversationID string, memberIDs, adminIDs []string) (*chatxdk.PreparedConversationChange, error) {
	return c.chat.PrepareGroupCreate(chatxdk.GroupCreateParams{
		PublicKeys:     publicKeys,
		ConversationID: conversationID,
		MemberIDs:      memberIDs,
		AdminIDs:       adminIDs,
	})
}

// PrepareGroupMembersChange prepares a member add under the session
// identity: rotated key + the two required signatures.
func (c *ChatCore) PrepareGroupMembersChange(publicKeys []chatxdk.PublicKeyInput, conversationID string, newMemberIDs, currentMemberIDs, currentAdminIDs []string) (*chatxdk.PreparedConversationChange, error) {
	return c.chat.PrepareGroupMembersChange(chatxdk.GroupMembersChangeParams{
		PublicKeys:              publicKeys,
		ConversationID:          conversationID,
		NewMemberIDs:            newMemberIDs,
		CurrentMemberIDs:        currentMemberIDs,
		CurrentAdminIDs:         currentAdminIDs,
		CurrentPendingMemberIDs: []string{},
	})
}

// -- Media streaming ----------------------------------------------------------

// mediaChunk is the fixed chunk size fed through the incremental stream API.
const mediaChunk = 1024 * 1024

// EncryptMedia encrypts a media blob with the incremental stream API.
//
// Feeding fixed-size chunks through Push keeps memory bounded no matter how
// large the file is; Finish emits the final frame that seals the stream
// (decryption fails without it).
func (c *ChatCore) EncryptMedia(plaintext, conversationKey []byte) ([]byte, error) {
	enc, err := c.chat.StreamEncryptor(conversationKey)
	if err != nil {
		return nil, err
	}
	defer enc.Close()
	var out []byte
	for offset := 0; offset < len(plaintext); offset += mediaChunk {
		end := min(offset+mediaChunk, len(plaintext))
		part, err := enc.Push(plaintext[offset:end])
		if err != nil {
			return nil, err
		}
		out = append(out, part...)
	}
	final, err := enc.Finish()
	if err != nil {
		return nil, err
	}
	return append(out, final...), nil
}

// DecryptMedia decrypts a media blob with the incremental stream API.
//
// Finish errors if the stream was truncated, so plaintext from Push must not
// be treated as complete until it succeeds.
func (c *ChatCore) DecryptMedia(ciphertext, conversationKey []byte) ([]byte, error) {
	dec, err := c.chat.StreamDecryptor(conversationKey)
	if err != nil {
		return nil, err
	}
	defer dec.Close()
	var out []byte
	for offset := 0; offset < len(ciphertext); offset += mediaChunk {
		end := min(offset+mediaChunk, len(ciphertext))
		part, err := dec.Push(ciphertext[offset:end])
		if err != nil {
			return nil, err
		}
		out = append(out, part...)
	}
	final, err := dec.Finish()
	if err != nil {
		return nil, err
	}
	return append(out, final...), nil
}

// -- Generic helpers (handy for metadata + tests) ---------------------------

// Encrypt encrypts a UTF-8 string with the raw 32-byte conversation key.
func (c *ChatCore) Encrypt(plaintext string, conversationKey []byte) (string, error) {
	return c.chat.Encrypt(plaintext, conversationKey)
}

// Decrypt decrypts base64 ciphertext to a UTF-8 string.
func (c *ChatCore) Decrypt(ciphertextB64 string, conversationKey []byte) (string, error) {
	return c.chat.Decrypt(ciphertextB64, conversationKey)
}

// MessageText returns the plain text of a decrypted Message event, or "".
func MessageText(event *chatxdk.Event) string {
	msg := event.AsMessage()
	if msg == nil {
		return ""
	}
	return msg.Text()
}

// PrepToRequest maps a prepared conversation change into the X API request shape.
//
// Works for 1:1 key changes (one signature) and group create / member add
// (two signatures). signingPublicKey is the sender's own signing key, which
// the API expects alongside each signature.
func PrepToRequest(prep *chatxdk.PreparedConversationChange, signingPublicKey string) map[string]any {
	participantKeys := make([]map[string]any, 0, len(prep.ParticipantKeys))
	for _, pk := range prep.ParticipantKeys {
		participantKeys = append(participantKeys, map[string]any{
			"user_id":                    pk.UserID,
			"encrypted_conversation_key": pk.EncryptedKey,
			"public_key_version":         pk.PublicKeyVersion,
		})
	}
	actionSignatures := make([]map[string]any, 0, len(prep.ActionSignatures))
	for _, sig := range prep.ActionSignatures {
		entry := map[string]any{
			"message_id":                   sig.MessageID,
			"encoded_message_event_detail": sig.EncodedMessageEventDetail,
			"message_event_signature": map[string]any{
				"signature":          sig.Signature,
				"signature_version":  sig.SignatureVersion,
				"public_key_version": sig.PublicKeyVersion,
				"signing_public_key": signingPublicKey,
			},
		}
		if sig.SignaturePayload != "" {
			entry["signature_payload"] = sig.SignaturePayload
		}
		actionSignatures = append(actionSignatures, entry)
	}
	return map[string]any{
		"conversation_key_version":      prep.ConversationKeyVersion,
		"conversation_participant_keys": participantKeys,
		"action_signatures":             actionSignatures,
	}
}

File: examples/rust/src/chat_core.rs
//! Crypto core for the Rust chat-xdk example bot.
//!
//! A thin, network-free wrapper around the `chat_xdk_core` binding.
//! Everything that touches the SDK lives here so it can be unit-tested directly
//! (see `tests/chat_core.rs`). The four core feature touchpoints are all here:
//!
//! - key management     -> [`ChatCore::load_keys`] / [`ChatCore::generate_and_register`]
//! - conversation keys  -> [`ChatCore::prepare_conversation_key_change`] / [`ChatCore::decrypt_conversation_key`]
//! - message encryption -> [`ChatCore::encrypt_reply`]
//! - event decryption   -> [`ChatCore::decrypt_batch`] (decrypt_events) and
//!   [`ChatCore::decrypt_one`] (decrypt_event)

use std::collections::HashMap;

use base64::{engine::general_purpose::STANDARD as B64, Engine};
use chat_xdk_core::crypto::keys::XChatConversationKey;
use chat_xdk_core::error::SdkError;
use chat_xdk_core::{
    AttachmentDescriptor, ChatCore as SdkCore, ConversationKeyChangeParams, DecryptEventsResult,
    EncryptMessageParams, EncryptReactionParams, EncryptReplyParams, EntityDescriptor, Event,
    GroupCreateParams, GroupMembersChangeParams, PreparedConversationChange, PublicKeyInput,
    PublicKeyRegistrationPayload, PublicKeys, SigningKeyEntry,
};
use serde_json::{json, Value};

/// Optional extras for [`ChatCore::encrypt_reply`].
#[derive(Debug, Clone, Default)]
pub struct ReplyOptions {
    /// Base64 raw event of the message to thread the reply under; `None`
    /// sends a fresh (unthreaded) message. The SDK derives the reply preview
    /// from it and embeds it so recipients can validate the preview.
    pub reply_to_event: Option<String>,
    /// Base64 raw key-change events, needed when the original message was
    /// encrypted under a different key version than the reply.
    pub reply_to_ckces: Option<Vec<String>>,
    /// Rich-text entities (mentions, URLs, etc.) as byte ranges in the text.
    pub entities: Option<Vec<EntityDescriptor>>,
    /// Attachment descriptors (e.g. a media reference) to embed.
    pub attachments: Option<Vec<AttachmentDescriptor>>,
    /// Disappearing-message lifetime in milliseconds.
    pub ttl_msec: Option<i64>,
}

/// Fields the X API expects for an encrypted message send.
#[derive(Debug, Clone, serde::Serialize)]
pub struct SendBody {
    pub message_id: String,
    pub encoded_message_create_event: String,
    pub encoded_message_event_signature: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conversation_token: Option<String>,
}

/// Wraps a single unlocked `chat_xdk_core::ChatCore` for one bot identity.
pub struct ChatCore {
    inner: SdkCore,
    pub signing_key_version: String,
}

impl ChatCore {
    pub fn new() -> Self {
        Self {
            inner: SdkCore::new(),
            signing_key_version: "1".to_string(),
        }
    }

    // -- Key management -----------------------------------------------------

    /// Import an existing base64 private-key blob (identity[+signing]) and unlock.
    pub fn load_keys(
        &mut self,
        private_keys_b64: &str,
        signing_key_version: &str,
    ) -> Result<(), SdkError> {
        let bytes = B64
            .decode(private_keys_b64)
            .map_err(|e| SdkError::Parse(format!("invalid base64 private keys: {e}")))?;
        self.inner
            .import_keys_with_version(&bytes, signing_key_version)?;
        self.signing_key_version = signing_key_version.to_string();
        Ok(())
    }

    /// Generate a fresh identity. Returns the registration payload to POST to
    /// the X API plus the exported private blob (base64) to persist locally.
    pub fn generate_and_register(
        &self,
    ) -> Result<(PublicKeyRegistrationPayload, String), SdkError> {
        let payload = self.inner.generate_keypairs()?;
        let exported = self.inner.export_keys()?;
        Ok((payload, B64.encode(exported)))
    }

    pub fn public_keys(&self) -> Result<PublicKeys, SdkError> {
        self.inner.get_public_keys()
    }

    /// Record the version the X API assigned to the registered public key;
    /// [`ChatCore::set_identity`] passes it to the SDK together with the
    /// user id.
    pub fn set_registered_version(&mut self, version: &str) {
        self.signing_key_version = version.to_string();
    }

    /// Set the session identity once; every signed action then defaults its
    /// `sender_id` and `signing_key_version` to these values.
    pub fn set_identity(&self, user_id: &str) {
        self.inner.set_identity(user_id, &self.signing_key_version);
    }

    // -- Conversation keys --------------------------------------------------

    pub fn prepare_conversation_key_change(
        &self,
        public_keys: &[PublicKeyInput],
        conversation_id: Option<&str>,
    ) -> Result<PreparedConversationChange, SdkError> {
        let mut params = ConversationKeyChangeParams::new(public_keys.to_vec());
        params.conversation_id = conversation_id.map(String::from);
        self.inner.prepare_conversation_key_change(params)
    }

    pub fn decrypt_conversation_key(
        &self,
        encrypted_key_b64: &str,
    ) -> Result<XChatConversationKey, SdkError> {
        self.inner.decrypt_conversation_key(encrypted_key_b64)
    }

    // -- Decryption: the two paths -----------------------------------------

    /// Batch path — used on initial conversation load (extracts conversation
    /// keys from KeyChange events, then decrypts every message).
    pub fn decrypt_batch(
        &self,
        events: &[&str],
        signing_keys: &[SigningKeyEntry],
    ) -> DecryptEventsResult {
        self.inner.decrypt_events(events, signing_keys)
    }

    /// Single-event path — used for each new event after the initial load.
    pub fn decrypt_one(
        &self,
        event_b64: &str,
        conversation_keys: &HashMap<String, XChatConversationKey>,
        signing_keys: &[SigningKeyEntry],
    ) -> Result<Event, SdkError> {
        self.inner
            .decrypt_event(event_b64, conversation_keys, signing_keys)
    }

    // -- Message encryption -------------------------------------------------

    /// Encrypt + sign a message, returning fields ready for the X API send.
    ///
    /// Without `reply_to_event` this sends a fresh message via
    /// `encrypt_message`; with it, the SDK's `encrypt_reply` builds a
    /// *threaded* reply whose preview is derived from that raw event.
    /// `entities` are byte ranges into the text; `attachments` are attachment
    /// descriptors (e.g. a media reference); `ttl_msec` makes the message
    /// disappear after the given lifetime. The sender identity comes from
    /// [`ChatCore::set_identity`].
    pub fn encrypt_reply(
        &self,
        conversation_id: &str,
        text: &str,
        conversation_key: &XChatConversationKey,
        conversation_key_version: &str,
        options: ReplyOptions,
    ) -> Result<SendBody, SdkError> {
        let payload = match &options.reply_to_event {
            None => {
                let mut params = EncryptMessageParams::new(conversation_id, text)
                    .with_conversation_key(conversation_key.to_bytes(), conversation_key_version);
                params.entities = options.entities;
                params.attachments = options.attachments;
                params.ttl_msec = options.ttl_msec;
                self.inner.encrypt_message(params)?
            }
            Some(reply_to_event) => {
                let mut params =
                    EncryptReplyParams::new(conversation_id, text, reply_to_event.clone())
                        .with_conversation_key(
                            conversation_key.to_bytes(),
                            conversation_key_version,
                        );
                params.reply_to_ckces = options.reply_to_ckces.clone();
                params.entities = options.entities;
                params.attachments = options.attachments;
                params.ttl_msec = options.ttl_msec;
                self.inner.encrypt_reply(params)?
            }
        };
        Ok(SendBody {
            // The SDK generates the message id and returns it in the payload.
            message_id: payload.message_id,
            encoded_message_create_event: payload.encrypted_content,
            encoded_message_event_signature: payload.encoded_event_signature,
            conversation_token: None,
        })
    }

    /// Encrypt + sign a reaction add/remove targeting a raw event: the SDK
    /// derives the conversation id and target sequence id from it.
    pub fn encrypt_reaction(
        &self,
        add: bool,
        target_event: &str,
        emoji: &str,
        conversation_key: &XChatConversationKey,
        conversation_key_version: &str,
    ) -> Result<SendBody, SdkError> {
        let params = EncryptReactionParams::new(target_event, emoji)
            .with_conversation_key(conversation_key.to_bytes(), conversation_key_version);
        let payload = if add {
            self.inner.encrypt_add_reaction(&params)?
        } else {
            self.inner.encrypt_remove_reaction(&params)?
        };
        Ok(SendBody {
            // The SDK generates the message id and returns it in the payload.
            message_id: payload.message_id,
            encoded_message_create_event: payload.encrypted_content,
            encoded_message_event_signature: payload.encoded_event_signature,
            conversation_token: None,
        })
    }

    // -- Group management -----------------------------------------------------

    /// Prepare a group creation: fresh key + the two required signatures.
    pub fn prepare_group_create(
        &self,
        public_keys: &[PublicKeyInput],
        conversation_id: &str,
        member_ids: &[String],
        admin_ids: &[String],
    ) -> Result<PreparedConversationChange, SdkError> {
        self.inner.prepare_group_create(GroupCreateParams::new(
            public_keys.to_vec(),
            conversation_id,
            member_ids.to_vec(),
            admin_ids.to_vec(),
        ))
    }

    /// Prepare a member add: rotated key + the two required signatures.
    pub fn prepare_group_members_change(
        &self,
        public_keys: &[PublicKeyInput],
        conversation_id: &str,
        new_member_ids: &[String],
        current_member_ids: &[String],
        current_admin_ids: &[String],
    ) -> Result<PreparedConversationChange, SdkError> {
        self.inner
            .prepare_group_members_change(GroupMembersChangeParams::new(
                public_keys.to_vec(),
                conversation_id,
                new_member_ids.to_vec(),
                current_member_ids.to_vec(),
                current_admin_ids.to_vec(),
                vec![],
            ))
    }

    // -- Media streaming -----------------------------------------------------

    /// Chunk size fed through the incremental stream encryptor/decryptor.
    pub const MEDIA_CHUNK: usize = 1024 * 1024;

    /// Encrypt a media blob with the incremental stream API.
    ///
    /// Feeding fixed-size chunks through `push` keeps memory bounded no
    /// matter how large the file is; `finish` emits the final frame that
    /// seals the stream (decryption fails without it).
    pub fn encrypt_media(
        &self,
        plaintext: &[u8],
        conversation_key: &XChatConversationKey,
    ) -> Result<Vec<u8>, SdkError> {
        let mut enc = self.inner.stream_encryptor(conversation_key)?;
        let mut out = Vec::new();
        for chunk in plaintext.chunks(Self::MEDIA_CHUNK) {
            out.extend(enc.push(chunk)?);
        }
        out.extend(enc.finish()?);
        Ok(out)
    }

    /// Decrypt a media blob with the incremental stream API.
    ///
    /// `finish` errors if the stream was truncated, so plaintext from `push`
    /// must not be treated as complete until it succeeds.
    pub fn decrypt_media(
        &self,
        ciphertext: &[u8],
        conversation_key: &XChatConversationKey,
    ) -> Result<Vec<u8>, SdkError> {
        let mut dec = self.inner.stream_decryptor(conversation_key)?;
        let mut out = Vec::new();
        for chunk in ciphertext.chunks(Self::MEDIA_CHUNK) {
            out.extend(dec.push(chunk)?);
        }
        out.extend(dec.finish()?);
        Ok(out)
    }

    // -- Generic helpers (handy for metadata + tests) -----------------------

    pub fn encrypt(
        &self,
        plaintext: &str,
        conversation_key: &XChatConversationKey,
    ) -> Result<String, SdkError> {
        self.inner.encrypt(plaintext, conversation_key)
    }

    pub fn decrypt(
        &self,
        ciphertext_b64: &str,
        conversation_key: &XChatConversationKey,
    ) -> Result<String, SdkError> {
        self.inner.decrypt(ciphertext_b64, conversation_key)
    }
}

impl Default for ChatCore {
    fn default() -> Self {
        Self::new()
    }
}

/// Pull the plain text out of a decrypted Message event, or `None`.
pub fn message_text(event: &Event) -> Option<&str> {
    match event {
        Event::Message(msg) => msg.text(),
        _ => None,
    }
}

/// Map a prepared conversation change into the X API request shape.
///
/// Works for 1:1 key changes (one signature) and group create / member add
/// (two signatures). `signing_public_key` is the sender's own signing key,
/// which the API expects alongside each signature.
pub fn prep_to_request(prep: &PreparedConversationChange, signing_public_key: &str) -> Value {
    let participant_keys: Vec<Value> = prep
        .participant_keys
        .iter()
        .map(|pk| {
            json!({
                "user_id": pk.user_id,
                "encrypted_conversation_key": pk.encrypted_key,
                "public_key_version": pk.public_key_version,
            })
        })
        .collect();
    let action_signatures: Vec<Value> = prep
        .action_signatures
        .iter()
        .map(|sig| {
            let mut entry = json!({
                "message_id": sig.message_id,
                "encoded_message_event_detail": sig.encoded_message_event_detail,
                "message_event_signature": {
                    "signature": sig.signature,
                    "signature_version": sig.signature_version,
                    "public_key_version": sig.public_key_version,
                    "signing_public_key": signing_public_key,
                },
            });
            // Conversation-key-change payloads are withheld by the SDK (they
            // embed the plaintext key); only forward one when it is present.
            if !sig.signature_payload.is_empty() {
                entry["signature_payload"] = json!(sig.signature_payload);
            }
            entry
        })
        .collect();
    json!({
        "conversation_key_version": prep.conversation_key_version,
        "conversation_participant_keys": participant_keys,
        "action_signatures": action_signatures,
    })
}
